sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/PushProjectionThroughLimitSuite.scala

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.optimizer
+
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.catalyst.dsl.expressions._
+import org.apache.spark.sql.catalyst.dsl.plans._
+import org.apache.spark.sql.catalyst.plans.PlanTest
+import org.apache.spark.sql.catalyst.plans.logical.{LocalRelation, LogicalPlan}
+import org.apache.spark.sql.catalyst.rules.RuleExecutor
+
+class PushProjectionThroughLimitSuite extends PlanTest {
+  object Optimize extends RuleExecutor[LogicalPlan] {
+    val batches = Batch("Optimizer Batch",
+      FixedPoint(100),
+      PushProjectionThroughLimit,
+      EliminateLimits) :: Nil
+  }
+
+  test("SPARK-40501: push projection through limit") {
+    val testRelation = LocalRelation.fromExternalRows(
+      Seq("a".attr.int, "b".attr.int, "c".attr.int),
+      1.to(20).map(_ => Row(1, 2, 3)))
+
+    val query1 = testRelation
+      .limit(10)
+      .select('a, 'b, 'c')
+      .limit(15).analyze
+    val optimized1 = Optimize.execute(query1)
+    val expected1 = testRelation
+      .select('a, 'b, 'c')
+      .limit(10).analyze
```

Comment:
I am curious why expected limit is 10 not 15.  `LIMIT 15` should be the top node.

I guess this is a question of how Spark coalesce adjacent limits.  How does Spark choose the value over a, b, c

LIMIT a
 --- LIMIT b
 --------LIMIT c
 ?

Reply:
Could you add a bit more complex test to include multiple LIMIT if that makes sense?

Reply:
> I am curious why expected limit is 10 not 15. `LIMIT 15` should be the top node.
> 
> I guess this is a question of how Spark coalesce adjacent limits. How does Spark choose the value over a, b, c
> 
> LIMIT a --- LIMIT b --------LIMIT c ?

EliminateLimits will be performed for optimization.

[https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala#L1977-L1985](url)
<img width="656" alt="image" src="https://user-images.githubusercontent.com/15246973/192070851-db7a2d1a-a250-4940-aedd-759f1bcab54e.png">

Reply:
> Could you add a bit more complex tests to include multiple LIMIT if that makes sense?
> 
> For example
> 
> ```
>     val query1 = testRelation
>       .limit(10)
>       .select('a, 'b, 'c')
>       .limit(20)
>       .select(`a`)
>       .limit(15)
>       .analyze
> ```

Done.

The UT only include rule: 
<img width="449" alt="image" src="https://user-images.githubusercontent.com/15246973/192071481-f06191df-4f20-42ec-ae3c-64961334fc39.png">